proposals/p1964.md

```diff
@@ -0,0 +1,140 @@
+# Character literals
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/1964)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+-   [Details](#details)
+    -   [Types](#types)
+    -   [Encoding](#encoding)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+
+<!-- tocstop -->
+
+## Problem
+
+This proposal specifies lexical rules for constant characters in Carbon.
+
+## Background
+
+We wish to provide a distinct lexical syntax for character literals versus
+string literals.
+
+In theory we could just reuse string literals for the purpose of character
+literals. However, it could benefit the readablity of our code if we had a
+distinct lexical syntax for character literals versus string literals.
+
+## Proposal
+
+The idea is to create and manage a character literal the same we would as a
+string, but using the single quote delimiter (') compared to the string double
+quote (").
+
+As with string literals, each character literal would have a different type.
+
+    var w: Character = 'w';
+
+We will not support:
+
+-   Multi-line literals
+-   "raw" literals (using #'x'#)
+-   Empty character literals (''')
+
+## Details
+
+A character literal is a sequence enclosed with single quotes delimiter ('),
+excluding:
+
+-   New line
+-   Single quote (`'`)
+-   Back-slash (`\`)
+-   Escape sequences
+
+The type of a character literal will depend on the the contents, so that `'c'`
+and `'b'`. However any `'\n'` and `'\u{A}'` would be of the same type (As when
+they are encoded, they are the same unicode entities `%0A`).
+
+### Types
```

Comment:
Could you explain this a bit more? Sorry, I'm not completely sure how this would look

Reply:
Look at the numeric literals proposals ( #144 ) and PRs ( #2092 ) to see how types encode values there, and how this can be expressed more concretely. The specifics would be different here, representing the character literal as a sequence consisting of elements that are each either a unicode code-point or a hex-encoded byte value.

An example of "decide whether the particular literal is represented in the variable's type by only looking at the types" would be explaining how we would allow:
```
let allowed: u8 = 'a';
```
but not allow:
```
let error1: u8 = '😃';
let error2: u8 = 'AB';
```
because the type of `'a'` indicates that it can be converted to `u8`, but the types of `'😃'` and `'AB'` indicate that they cannot.